### ISSUE TITLE
systemd-user: pam_keyinit(systemd-user:session): Unable to change GID temporarily

### ISSUE THREAD

**Issue by yuwata** [systemd/systemd#4238]
### Submission type
- [x] Bug report
- [ ] Request for enhancement (RFE)

_NOTE: Do not submit anything other than bug reports or RFEs via the issue tracker!_
### systemd version the issue has been seen with

> Git snapshot 95cbb83c2061237bfd3760c470f6c91cf3ec069b.
> Note that e4662e553ccc7abccb80bf74eb79544c2e26dabe does not have this issue.

_NOTE: Do not submit bug reports about anything but the two most recently released systemd versions upstream!_
### Used distribution

> fedora 24 (x86_64) with the above systemd git snapshot.
### In case of bug report: Expected behaviour you didn't see

> Users can exit from their sessions without the error message below.
### In case of bug report: Unexpected behaviour you saw

> When logout from the user session, the error message "pam_keyinit(systemd-user:session): Unable to change GID to (GID of the user) temporarily" is shown.
> 
> ```
> sshd[757]: Received disconnect from 192.168.0.12 port 51830:11: disconnected by user
> sshd[757]: Disconnected from 192.168.0.12 port 51830
> sshd[732]: pam_unix(sshd:session): session closed for user yu.watanabe
> systemd-logind[600]: Removed session 3.
> systemd[1]: Stopping User Manager for UID 1014...
> systemd[734]: Stopped target Default.
> systemd[734]: Stopped target Basic System.
> systemd[734]: Stopped target Sockets.
> systemd[734]: Closed D-Bus User Message Bus Socket.
> systemd[734]: Reached target Shutdown.
> systemd[734]: Starting Exit the Session...
> systemd[734]: Stopped target Paths.
> systemd[734]: Stopped target Timers.
> systemd[734]: Received SIGRTMIN+24 from PID 825 (kill).
> systemd[744]: pam_keyinit(systemd-user:session): Unable to change GID to 1000 temporarily
> systemd[1]: Stopped User Manager for UID 1014.
> systemd[1]: Removed slice User Slice of yu.watanabe.
> ```
### In case of bug report: Steps to reproduce the problem

> ssh login or graphical login via lightdm by user (not root) and logout the session.

**Comment by tixxdz** [comment]
Referencing this https://github.com/yuwata/systemd/commit/4372c5521ce9df38a85ca2c70f72f3f108536590#commitcomment-19296428  here for now.

**Comment by evverx** [comment]
Hm, indeed
- master

``` sh
-bash-4.3$ ps -C '(sd-pam)'
  PID TTY          TIME CMD
  260 ?        00:00:00 (sd-pam)

-bash-4.3$ grep [UG]id /proc/260/status
Uid:    1002    1002    1002    1002
Gid:    0       0       0       0

...logout...

-bash-4.3# journalctl -b _PID=260 --no-pager --no-hostname
-- Logs begin at Wed 2016-10-05 08:21:16 UTC, end at Wed 2016-10-05 09:33:07 UTC. --
Oct 05 09:31:40 systemd[260]: pam_keyinit(systemd-user:session): CLOSE 1,1061645500,1
Oct 05 09:31:40 systemd[260]: pam_keyinit(systemd-user:session): REVOKE 1061645500
Oct 05 09:31:40 systemd[260]: pam_keyinit(systemd-user:session): UID:1002 [1002]  GID:1002 [0]
Oct 05 09:31:40 systemd[260]: pam_keyinit(systemd-user:session): Unable to change GID to 1002 temporarily
Oct 05 09:31:40 systemd[260]: pam_unix(systemd-user:session): session closed for user hola
```
- e4662e5 

``` sh
-bash-4.3$ ps -C '(sd-pam)'
  PID TTY          TIME CMD
  259 ?        00:00:00 (sd-pam)

-bash-4.3$ grep [UG]id /proc/259/status
Uid:    1002    1002    1002    1002
Gid:    1002    1002    1002    1002

...logout...
-bash-4.3# journalctl -b _PID=259 --no-pager --no-hostname
-- Logs begin at Wed 2016-10-05 09:19:50 UTC, end at Wed 2016-10-05 09:37:50 UTC. --
Oct 05 09:37:02 systemd[259]: pam_keyinit(systemd-user:session): CLOSE 1,340746107,1
Oct 05 09:37:02 systemd[259]: pam_keyinit(systemd-user:session): REVOKE 340746107
Oct 05 09:37:02 systemd[259]: pam_keyinit(systemd-user:session): UID:1002 [1002]  GID:1002 [1002]
Oct 05 09:37:02 systemd[259]: pam_unix(systemd-user:session): session closed for user hola
```

**Comment by evverx** [comment]
096424d1230e0a0339735c51b43949809e972430 is the first bad commit

**Comment by poettering** [comment]
Hmm, so I figure the PAM stub process we keep around for invoking the PAM shutdown hooks is at fault here. We keep it around with dropped UID, but with the original GID... I figure we should also fix up the GID set for it now.

Background: previously we'd set up the GID early, before doing PAM, but the UID after doing so. 096424d changed that. Now we set up PAM first, and the UID+GID later. However, for the PAM stub we manually fixed up the UID then, but still not the GID, and I think that's something we need to fix. The fix should be pretty simple.
